src/mono/mono/metadata/components.c

```diff
@@ -160,6 +155,12 @@ try_load (const char* dir, const MonoComponentEntry *component, const char* comp
 static MonoComponentInitFn
 load_component (const MonoComponentEntry *component, MonoDl **lib_out)
 {
+	// If init method has been static linked not using stub library, use that instead of dynamic component.
+	if (component->init() && component->init()->available ()) {
```

Comment:
I wonder if we should pass `MONO_DL_LOCAL` to `mono_dl_open` explicitly in `try_load`.  (It's the default, but maybe it's better to be explicit)

---

the rest of this comment is just me trying to understand how this all works.

So if I understand what's going on:
1. suppose we build components dynamically, and link the stubs statically into libmonosgen-2.0
2. We now have the symbol `mono_component_event_pipe_init` in two places:
   a. in libmonosgen-2.0 coming from the component stub,
   b. in libmono-component-event_pipe

So initially we call `component->init()` and get the stub vtable in which `available()` returns false and we probe for the dynamic component.  Then we dlopen the component library, and dlsym `mono_component_event_pipe_init` and call that function.

So in theory we have two copies of that symbol floating around now, but one of them is local, but that's ok.

Reply:
Yes, so the symbol in libmonosgen-2.0.so is local and should not be visible outside or added to the global symbol table. The code in components.c should be a static link the stub symbol when building components dynamically in libmonosgen-2.0.so and when we build libmonosgen-2.0.a its undefined and will be linked to either real init method or stub version, depending on how user builds final binary.

Reply:
For the static components scenario all should remain the same as before.